File: Arrays/kadanes-algorithm/maximum-subarray-sum.py

#  Brute - Force Approach
def maximumSubArraySum(nums):
    n = len(nums)
    maxSum = float('-inf')
    for start in range(n):
        for end in range(start, n):
            currSum = 0
            for i in range(start, end+1):
                currSum += nums[i]
                maxSum = max(maxSum, currSum)
                
    print(maxSum) #  9
    
#  Optimal Approach
def maximumSubArraySumOptimal(nums):
    maxSum = float('-inf')
    currSum = 0
    n = len(nums)
    for i in range(n):
        currSum += nums[i]
        maxSum = max(currSum, maxSum)
        if currSum < 0:
            currSum = 0
    print(maxSum) # 9

def main():
    nums = [5, -3, 2, 5, -5, 3]
    maximumSubArraySum(nums)
    maximumSubArraySumOptimal(nums)
    
main()
    

File: Arrays/kadanes-algorithm/maximum-subarray-sum.c
#include<stdio.h>
#include<limits.h>
#include<math.h>


// Brute - Force Approach
void maximumSubArraySum(int nums[], int n) {
    int maxSum = INT_MIN;
    for(int start = 0; start < n; start++) {
        for(int end = start; end < n; end++) {
            int currSum = 0;
            for(int i = start; i <= end; i++) {
                currSum += nums[i];
                if(currSum > maxSum) {
                    maxSum = currSum;
                }
            }
        }
    }
    printf("%d\n", maxSum); // 17
}


// Optimal approach
void maximumSubArraySumOptimal(int nums[], int n) {
    int maxSum = INT_MIN;
    int currSum = 0;
    for(int i = 0; i < n; i++) {
        currSum += nums[i];
        if(currSum > maxSum) {
            maxSum = currSum;
        }
        if(currSum < 0) {
            currSum = 0;
        }
    }
    printf("%d\n", maxSum); // 17
}


int main() {
    int nums[] = {1, 4, -3, 5, -2, 7, 5};
    int n = sizeof(nums) / sizeof(nums[0]);
    maximumSubArraySum(nums, n);
    maximumSubArraySumOptimal(nums, n);
}